streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/ForeignJoinSubscriptionSendProcessorSupplier.java

```diff
@@ -52,35 +61,38 @@ public KTableRepartitionerProcessorSupplier(final ValueMapper<V, KO> extractor,
 
     private class UnbindChangeProcessor extends AbstractProcessor<K, Change<V>> {
         
+        @SuppressWarnings("unchecked")
         @Override
         public void init(final ProcessorContext context) {
             super.init(context);
+            // get default key serde if it wasn't supplied directly at construction
+            if (foreignKeySerializer == null) {
+                foreignKeySerializer = (Serializer<KO>) context.keySerde().serializer();
+            }
         }
 
         @Override
         public void process(final K key, final Change<V> change) {
             final long[] currentHash = change.newValue == null ?
                     null :
-                    Murmur3.hash128(valueSerializer.serialize(context().topic(), change.newValue));
+                    Murmur3.hash128(valueSerializer.serialize(repartitionTopicName, change.newValue));
 
             if (change.oldValue != null) {
                 final KO oldForeignKey = mapper.apply(change.oldValue);
                 if (change.newValue != null) {
                     final KO newForeignKey = mapper.apply(change.newValue);
 
-                    //Requires equal to be defined...
-                    if (oldForeignKey.equals(newForeignKey)) {
-                        //Same foreign key. Just propagate onwards.
-                        context().forward(newForeignKey, new SubscriptionWrapper<>(currentHash, PROPAGATE_NULL_IF_NO_FK_VAL_AVAILABLE, key));
-                    } else {
+                    final byte[] serialOldForeignKey = foreignKeySerializer.serialize(repartitionTopicName, oldForeignKey);
+                    final byte[] serialNewForeignKey = foreignKeySerializer.serialize(repartitionTopicName, newForeignKey);
+                    if (!Arrays.equals(serialNewForeignKey, serialOldForeignKey)) {
```

Comment:
Fixes the problem with not necessarily having equals defined on the FK (it's really important for performance that this check be airtight, otherwise every single record would cause an unsubscribe followed by a resubscribe).